### ISSUE TITLE
Composite artifact can't be resolved in buildSrc

### ISSUE THREAD

**Issue by uklance** [gradle/gradle#3768]
I have a composite build with a plugin and a consumer of the plugin. 

- When I declare the plugin artifact in `buildscript.dependencies.classpath` in `plugin-consumer/build.gradle` the plugin is correctly resolved via the composite. 
- When I declare the plugin artifact in `dependencies.compile` in `plugin-consumer/buildSrc/build.gradle` the plugin can't be found

Please see [this project](https://github.com/uklance/gradle-composite-bug)

1. [this commit](https://github.com/uklance/gradle-composite-bug/commit/d3879dc813c7fe52c885be4b680b4301542a826a) works where ConcreteTask extends AbstractTask in `plugin-consumer/build.gradle`
2. [this commit](https://github.com/uklance/gradle-composite-bug/commit/6447778ed2ff72430040025eaf6002082cd5d548) fails where I attempt to move `ConcreteTask` to `plugin-consumer/buildSrc/`

Error message
```
FAILURE: Build failed with an exception.

* Where:
Settings file 'C:\code\gradle-composite-bug\settings.gradle' line: 2

* What went wrong:
Could not resolve all files for configuration ':plugin-consumer:buildSrc:runtimeClasspath'.
> Could not find com.example:my-plugin:1.0-SNAPSHOT.
  Searched in the following locations:
      https://repo1.maven.org/maven2/com/example/my-plugin/1.0-SNAPSHOT/maven-metadata.xml
      https://repo1.maven.org/maven2/com/example/my-plugin/1.0-SNAPSHOT/my-plugin-1.0-SNAPSHOT.pom
      https://repo1.maven.org/maven2/com/example/my-plugin/1.0-SNAPSHOT/my-plugin-1.0-SNAPSHOT.jar
  Required by:
      project :buildSrc
```

### Expected Behavior
In a composite build, a plugin referenced in `buildSrc/build.gradle` should be resolved 

### Current Behavior
In a composite build, a plugin referenced in `buildSrc/build.gradle` cannot be resolved 

### Context
I'm trying to build, consume and extend a plugin in a single composite build but as soon as I try to move logic to `buildSrc` the build is failing

### Steps to Reproduce (for bugs)
Clone the repository at https://github.com/uklance/gradle-composite-bug.git
Run `gradle build` from the root directory

### Your Environment
------------------------------------------------------------
Gradle 4.1
------------------------------------------------------------

Build time:   2017-08-07 14:38:48 UTC
Revision:     941559e020f6c357ebb08d5c67acdb858a3defc2

Groovy:       2.4.11
Ant:          Apache Ant(TM) version 1.9.6 compiled on June 29 2015
JVM:          1.8.0_45 (Oracle Corporation 25.45-b02)
OS:           Windows 7 6.1 amd64

**Comment by staffanf** [comment]
@oehme I followed the advice in https://www.stefan-oehme.com/use-buildSrc to move my own plugins module dependencies to buildSrc and now I can't use --includeBuild to get the fast feedback. This removed the need for some boilerplate `buildScript` blocks which I had in several build scripts .

We use the plugins both in buildSrc for extending tasks for specific purposes and in the subprojects as-is.
If I put a `buildScript.dependencies` block in the subprojects then the included build will be pulled in and built but the buildSrc classes seems to be first in the classpath, effectively killing that workaround.

**Comment by oehme** [comment]
buildSrc is not yet a normal composite and thus can't take part in substitution. This is the long term goal though. 

Instead of using buildSrc you can use an included build that is part of your repo and always included. Then just depend on it in your root project buildscript block. This roughly reproduces the behavior of buildSrc, but should also make including other builds work. I might write another post about that. Each has its limitations at the moment, but we'll get to a unified model eventually :)

**Comment by svermeulen** [comment]
I am finding that replacing `buildSrc` with an included build that is always included has its own set of problems.

Given a project A, with an included build B (that functions similarly to buildSrc), any other projects cannot simply `includeBuild A` without also adding `includeBuild B`, since you cannot currently have composite builds that include other composite builds.  In other words, all included builds need to be declared in the root project instead of any included projects.

Once buildSrc becomes a normal included build I expect this will go away though, so looking forward to that

**Comment by dsilvasc** [comment]
@svermeulen Doesn't gradle support nested composite builds as of 4.10?
https://docs.gradle.org/4.10/release-notes.html#nested-included-builds

**Comment by svermeulen** [comment]
@dsilvasc You're right, my issue is more complicated than that.  It seems that nested composite builds do work except in certain cases.  In my case, they don't work when the included build includes another build for use inside the buildscript {} block.  In this case, the nested included build needs to be declared in the root project.  I made a minimal example here showing the issue:

https://github.com/svermeulen/GradleCompositeBuildBugExample

To repro, just comment out the line `includeBuild '../bar'` in `qux/settings.gradle` and then, inside the `qux` directory, run `./gradlew foo`.  That line is necessary in order for that task to work, when really it shouldn't be necessary if nested composite builds work

**Comment by jjohannes** [comment]
The original issue should be fixed in `6.7` (release candidate is available now).

See: https://docs.gradle.org/6.7-rc-1/release-notes.html#build-src

(For other problems still present in 6.7, please open a separate issue)
